internal/argocd/health.go

```diff
@@ -0,0 +1,278 @@
+package argocd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path"
+	"slices"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	argocd "github.com/akuity/kargo/internal/controller/argocd/api/v1alpha1"
+	"github.com/akuity/kargo/internal/git"
+)
+
+// healthErrorConditions are the v1alpha1.ApplicationConditionType conditions
+// that indicate an Argo CD Application is unhealthy.
+var healthErrorConditions = []argocd.ApplicationConditionType{
+	argocd.ApplicationConditionComparisonError,
+	argocd.ApplicationConditionInvalidSpecError,
+}
+
+// compositeError is an interface for wrapped standard errors produced by
+// errors.Join.
+type compositeError interface {
+	// Unwrap returns the wrapped errors.
+	Unwrap() []error
+}
+
+// ApplicationHealth is a health evaluator for Argo CD Applications.
+type ApplicationHealth struct {
+	Client client.Client
+}
```

Comment:
Nit: This type is a teensy bit brittle because it requires callers outside this package to have knowledge of how to initialize it. (Not that this isn't a pervasive problem in Go due to its lack of actual constructors.)

I would have a slight preference that we do this:

```
type AppHealthEvaluator interface {
	EvaluateHealth(context.Context, kargoapi.FreightReference, []kargoapi.ArgoCDAppUpdate) *kargoapi.Health
}

type appHealthEvaluator struct {
	client client.Client
}

func NewAppHealthEvaluator(c client.Client) AppHealthEvaluator {
	return &appHealthEvaluator {
		client: c,
	}
}

// etc...
```

Reply:
I can live with this suggestion.

The reason for the current code is because I expected the caller to be aware of the structure — which is a constructor by itself. While I did not expect there to be multiple implementations from the side of the `argocd` package — which would be reason to define an interface on this side.

This approach aligns with Go's philosophy of favoring composition over inheritance, which promotes defining interfaces on the side of the user. However, this is also frowned upon by many (including myself at times), and I do agree that your approach may be the nicer way of doing it here.

Reply:
Moved things around in 3742ef807e93b86de49b15f33c91f52c5e100a08